### DISCUSSION TITLE
can you replace expressjs with fastify?

### DISCUSSION THREAD

**Discussion by anlexN** [directus/directus#15380] [Archive]
# **expressjs is not actively maintained.**

**Comment by LasseRosenow** [comment]
As the author didn't give any more information, I would add some. `Express` is really old. This performance is not so amazing. In fact the developers of `Express` are now focusing more on working on the new `Koa` Framework, which is kind of like the successor of `Express`.

Regarding `Fastify`: based on some online statistics it's around 28% faster than `Express`, which is a huge deal I think.

Statistics: https://facsiaginsa.com/nodejs/comparing-fastify-vs-express

So replacing `Express` with `Fastify` could really improve the performance of directus applications:)


Out of context:
I myself still thing the best thing would be for the Directus API to be written in Rust, as not only the performance gains would be beyond everything node could ever do and also very importantly the memory usage would decrease A LOT
Rust statistics: https://medium.com/@maxsparr0w/performance-of-node-js-compared-to-actix-web-37f20810fb1a

**Reply by anlexN** [reply]
I don't like koa. expressjs and koa are from the same team, just for a different branch. and koa is not in (OPENJSF)[https://openjsf.org/]. so It's a pity.
As for rust, the learning cost is very very high. oh my god. my friend.

**Reply by Nitwel** [reply]
If you can either train the whole Directus team to develop in Rust or hire 5 to 10 new Developers that have a lot of experience in Rust, this would be a welcomming change. But jokes aside, it is super nice to be able to work with the same language on both front and backend. We can share functionallity between both e.g. `@directus/shared` and object handling is completely the same as it's using the same underlying language. 

Besides that, in the case of a language shift it might be more beneficial to look at all possible languages as Rust might not be the best fit for Directus. I would definitely want to research into Go, Java Spring and Ruby on Rails as I heard great thing from all of those.

Especially Go sounds really interesting performance wise but I'm also excited to read through what others suggest and/or have experience with. 😄

**Reply by anlexN** [reply]
we should stick to javascript full stack, not other languages.

**Reply by LasseRosenow** [reply]
> As for rust, the learning cost is very very high. oh my god.

Actually the learning curve is not so high, at least in my experience. The first few days might be a little confusing, but then you get a hang if it really fast. For me it feels almost like writing modern java 😱

But I understand that you would prefer to share the stack between front- and backend.

**Reply by anlexN** [reply]
1. I took very much time to master JAVASCRIPT, HTML, CSS, POSTGRESQL
2. I think js full stack is a good fit for me.
3. ECMAScript is the industrial standard
4. I have no time to master other languages.
5. I need an active community.
6. SQL is the industrial standard, but NoSQL is not.
7. I am Chinese. I need free, open-source, standard things. although our company (HUAWEI) is sanctioned by the United States. our government does not sanction any global company.
8. we respect any global company and do not discriminate against any global one.
9. our country's inflation rate is not high like others. it is low. Not coming to China is your loss

**Reply by LasseRosenow** [reply]
What is the reason for this random post?
What is the connection between using fastify etc. And SQL? What does all this have to do that you need free open source standard things? Everything is open source in here. Why do you talk about if someone wants to come to china or not? And what is the reason for all this political stuff?!

A little bit awkward

**Reply by anlexN** [reply]
1. directus can be better
2. directus use some tools that are not actively maintained.
3. directus do not need to take time to adopt mongodb
4. directus cloud have no china location

**Reply by Nitwel** [reply]
Community cloud is US only as it is free and standard cloud covers US, EU and Asia so maybe you have overlooked that. 😄 

![grafik](https://user-images.githubusercontent.com/22577866/188274671-be701d93-2931-4d99-b821-5b65dcb14f6f.png)

**Reply by anlexN** [reply]
you have Asia locations, but no china mainland. I have reviewed your docs step by step, and I have communicated with your sales. I make sure.

**Comment by rijkvanzanten** [comment]
I'm closing this due to the original post not having any context or further elaboration[^1] and the main thread getting completely off-topic immediately. Happy to discuss individual points (like Rust vs JS, Fastify vs Express, SQL vs NoSQL) in separate discussions, but lets keep them on-topic, respectful, and project driven as per [the code of conduct](https://github.com/directus/directus/blob/main/code_of_conduct.md#our-standards) 🙂 

[^1]: Also the main point "not maintained" seems to be false as the express project sees regular updates https://github.com/expressjs/express 🤷🏻
